### ISSUE TITLE
HoM on Plutonia map11 with OpenGL

### ISSUE THREAD

**Issue by snoggie1487** [coelckers/prboom-plus#226]
There is a HoM bug at the very start of Plutonia map11. 
![doom02](https://user-images.githubusercontent.com/38010396/109367688-67e02800-785c-11eb-95d5-76e43b6aa6d9.png)
![doom03](https://user-images.githubusercontent.com/38010396/109367689-69115500-785c-11eb-87ee-8f9eb2031164.png)
![doom05](https://user-images.githubusercontent.com/38010396/109367691-69a9eb80-785c-11eb-96cb-5aa54128db15.png)
![doom00](https://user-images.githubusercontent.com/38010396/109367692-6a428200-785c-11eb-9a87-f7867fcc3ded.png)
![doom01](https://user-images.githubusercontent.com/38010396/109367693-6a428200-785c-11eb-994d-5bed32ebfffe.png)

**Comment by runlow** [comment]
Not sure if related and if it isn't the mapmaker's fault - but there's pretty bad HOM in Alien Vendetta map11 when you enter the waterfall.

![2021-02-28-093007_1920x1080_scrot](https://user-images.githubusercontent.com/2176657/109413906-bd751b80-79a7-11eb-9f94-b123a12cb276.png)

![2021-02-28-093000_1920x1080_scrot](https://user-images.githubusercontent.com/2176657/109413901-bb12c180-79a7-11eb-8589-5baf54b53d66.png)

**Comment by snoggie1487** [comment]
This occurs when a sector fully connects with the floor / ceiling (not both) of the surrounding sector, and has no sidedef textures. This was used in Plutonia map11 as an invisible monster-los blocking solid wall. I made a quick map to demonstrate this, try looking at the ground with software rendering and then opengl rendering. 
[opengl_bug.zip](https://github.com/coelckers/prboom-plus/files/6127298/opengl_bug.zip)

**Comment by fabiangreffrath** [comment]
Thank you very much for the test map! I hope this will help somebody nail down this issue down. Unfortunately, this somebody won't be me, I am OpenGL-illiterate. 😞 

However, it raises one question: If the sector's floor is above the view plane (`viewz`), you will be looking beneath the floor, so strictly speaking there can't be a visible floor flat on the ground. Which part of the software renderer decides to continue using the floor flat of the surrounding sector in this case?

**Comment by JadingTsunami** [comment]
I think PrBoom tries to fix this, but it seems like it's broken or maybe just not working in these cases.

The following flags appear to try and address this using stencils (which seems like the right way to fix it).

```c
#define GLDWF_TOPFLUD 5 //e6y: project the ceiling plane into the gap
#define GLDWF_BOTFLUD 6 //e6y: project the floor plane into the gap
```

I haven't had much time to look into these so if others know more please do comment.

**Comment by runlow** [comment]
Not sure if related or helpful - there's HOM in DOOM2.WAD on map21 above the rocket launcher in opengl mode (c170d6362c7 build, not the fork/branch).
![2021-03-20-144202_1920x1080_scrot](https://user-images.githubusercontent.com/2176657/111874001-0b6bb680-898b-11eb-9d42-67ef5ec72f94.png)

**Comment by runlow** [comment]
@JadingTsunami haven't delved much into the technical details yet and not sure this helps but I tested your branch:

```
PLUTONIA.WAD     map11   OK           in front of archviles
HR.WAD           map01   OK           above still pinkies
DOOM2.WAD        map21   still hom    above the rocket launcher
AV.WAD           map11   still hom    inside the waterfall

```

**Comment by JadingTsunami** [comment]
Yes, the DOOM2.WAD Map21 glitch is because the map is malformed. It looks like there was a connecting sector at one point between the triangle sector with the Rocket Launcher and the steps up to the next platform, but it was deleted at some point and the sector references were never updated. That would need a different approach to fix malformed maps; or possibly a custom sector merge function like what's done for E3M8 in some ports. Either way, I think it would be tackled separately. But it's a good find.

AV.WAD is also due to sector references. It looks like that tunnel uses self-referencing sectors in a unique way. It won't be fixed by the issue here and would be a separate issue.

Good to know the other two issues are fixed for you as well, though.

**Comment by fabiangreffrath** [comment]
@JadingTsunami Awesome, thank you for having a look at this!

**Comment by JadingTsunami** [comment]
I think some discussion will be needed on this one; I have explored two options:

The first option would be to try and create a stencil around the gap and fill it in with an adjacent flat at the same vertical level as the gap itself. This is more complicated than it seems, because at the point in the renderer where walls are rendered it's not so clear to me how to extract adjacent sector information in an efficient way.

The option I have implemented does two things, and is I think simpler, but it requires some backstory.

In glboom, before stenciling, there was a method to bleed-over "fake" sectors to emulate deep water/invisible ceiling software renderer effects. This was done by identifying sectors that have bleed-over and pairing them as floor/ceiling bleed candidates. There's an existing system to identify and catalog these fake sector candidates for each sector that will require bleed-over.

To fix the issue here, I re-enabled this old system in order to grab the candidate fake sectors during preprocessing. Then add fake flats matching the target sector candidate identified by the preprocessor, except using the lowest floor/ceiling adjacent to the invisible sector. This works almost all of the time, but will fail for odd cases like having a control sector adjacent to a dummy sector copy outside the map boundaries with a lower sector height than anything in the player's view. I've never seen this done personally but it's possible, and will cause a similar HOM effect with the sector "floating" lower.

I would like to hear thoughts on the approach or if there's a better way people had in mind to fix this.
